### ISSUE TITLE
Connection schema name is always public in task

### ISSUE THREAD

**Issue by faxioman** [maciej-gol/tenant-schemas-celery#12]
Hi.
I have configured my celery.py script using this syntax:

```
from __future__ import absolute_import, unicode_literals
import os
from tenant_schemas_celery.app import CeleryApp

os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'MYPROJ.settings')
app = CeleryApp('MYPROJ')

app.config_from_object('django.conf:settings', namespace='CELERY')


app.autodiscover_tasks()
```

But when a task is executed from a tenant, inside the task the tenant is alway public... Am I missing something?

tenant-schemas-celery==0.1.6
celery[redis]==4.0.2
django==1.9.7
django-tenant-schemas==1.6.2

Thanks

**Comment by alexjillard** [comment]
@faxioman I was looking into this and it seems like in some circumstances the schema name is stored in the task.request rather than task.request.headers. I made a small update that's fixed the issue for me. I'm still in the process of testing it though, but feel free to check it out and see if it works on your end.

Just a small alteration to the logic that tries to grab the schema name from the task:
https://github.com/alexjillard/tenant-schemas-celery/blob/celery-4-no-schema-bug/tenant_schemas_celery/app.py#L19

**Comment by maciej-gol** [comment]
@alexjillard do you know when this happens?

**Comment by alexjillard** [comment]
@maciej-gol We're able to cause this using chords.

We're using celery 4.0.2 and django 1.8.18

```python
def _fetch_and_parse_foo(self):
    tasks = self._chord_fetch_header()
    callback = _chord_parse_callback.s(
        self.__class__.__module__, 
        self.__class__.__name__,
        self.args,
        self.kwargs
    )
    results = chord(tasks)(callback)
    return results

def _chord_fetch_header(self):
    header = [
        self._bar_task().s(self.foo_id, self.tenant.pk, multiline=True).on_error(on_chord_error.s()),
        self._bar_task().s(self.foo_id, self.tenant.pk,).on_error(on_chord_error.s()),
    ]
    if self.baz_required:
        header.extend([
            self._baz_task().s(self.foo_id, self.tenant.pk, multiline=True).on_error(on_chord_error.s()),
            self._baz_task().s(self.foo_id, self.tenant.pk,).on_error(on_chord_error.s()),
        ])
    return header
```

**Comment by maciej-gol** [comment]
And the wrong schema name is in the original task, or in the callback?

**Comment by alexjillard** [comment]
The schema name was not preset in `task.request.headers` in the tasks. The chord callback never fired because all the tasks failed.

**Comment by maciej-gol** [comment]
Could you give more context? The following works fine (I thought it could be something with chords or groups):

```python
@app.task
def print_schema():
    print connection.schema_name


with tenant_context(tenant):
    chord([print_schema.s(), print_schema.s()])(print_schema.s())
```

**Comment by oplahcinski** [comment]
I just want to add that I also experienced this issue.

I have some production servers running tenant-schemas-celery==0.1.5 and they worked fine, when I updated to tenant-schemas-celery==0.1.6 I get the schema set to public.

The task is super simple
```

from proj.settings.celery import app
from proj.users.tenant_user.models import TenantUser


@app.task
def send_email(user_pk, uuid, func_name, **kwargs):
    user = TenantUser.objects.get(pk=user_pk, uuid=uuid)
    if hasattr(user, func_name):
        getattr(user, func_name)(**kwargs)

```

Im just using it to call a method on my tenant user that will send an email. The tenant user cant be found because the schema is public

relevant freeze:
```
tenant-schemas-celery==0.1.6
Django==1.8.14
django-tenant-schemas==1.6.2
celery==4.0.2
```

**Comment by maciej-gol** [comment]
@oplahcinski , thanks, will look into it.

19 kwi 2017 20:07 "oplahcinski" <notifications@github.com> napisał(a):

**Comment by maciej-gol** [comment]
@oplahcinski , are you by any chance using `Redis` or other broker than `RMQ`?

**Comment by faxioman** [comment]
@maciej-gol I'm using redis ... I'll try your fix.

**Comment by maciej-gol** [comment]
@faxioman any luck with it?

**Comment by maciej-gol** [comment]
Should've been fixed in `0.1.17`
